### PR TITLE
Allow usage of a custom `OptimizerChain` (#106)

### PR DESCRIPTION
**Pull request by nlemoine**
Sorry, I forgot that `array_filter` preserves indexes. This is now fixed.

### CODE REVIEW

File: src/Image.php [hunk 1]
```diff
@@ -163,11 +163,11 @@ protected function performOptimization($path, array $optimizerChainConfiguration
...
 
             $optimizers = array_map(function (array $optimizerOptions, string $optimizerClassName) use ($existingOptimizers) {
-                $optimizer = array_filter($existingOptimizers, function ($optimizer) use ($optimizerClassName) {
+                $optimizer = array_values(array_filter($existingOptimizers, function ($optimizer) use ($optimizerClassName) {
                     return get_class($optimizer) === $optimizerClassName;
-                });
+                }));
 
-                $optimizer = count($optimizer) > 0 && $optimizer[0] instanceof BaseOptimizer ? $optimizer[0] : new $optimizerClassName;
+                $optimizer = isset($optimizer[0]) && $optimizer[0] instanceof BaseOptimizer ? $optimizer[0] : new $optimizerClassName;
```
Comment (freekmurze) [inline src/Image.php:170]: Is this change necessary?

Comment (nlemoine) [reply src/Image.php:170]: It's safer.

```
php
$optimizer = [
    1 => (JpegOptim instance)
];
```

`count($optimizer) > 0` will be evaluated whereas `isset($optimizer[0])`. Even if it's not supposed to happen.

**Comment by freekmurze** [comment]
Thanks!

**Comment by 0xb4lint** [comment]
It seems this commit broke my image optimization jobs with the following error message:
`Process errored with 'sh: /jpegoptim: No such file or directory'`

In these PRs there is a new function call `->setBinaryPath($optimizer->binaryPath)`, which utilizes [this function](https://github.com/spatie/image-optimizer/blob/master/src/Optimizers/BaseOptimizer.php#L25-L34):

```php
abstract class BaseOptimizer implements Optimizer
{
    /* ... */
    public $binaryPath = '';

    /* ... */
    public function setBinaryPath(string $binaryPath)
    {
        if (substr($binaryPath, -1) !== DIRECTORY_SEPARATOR) {
            $binaryPath = $binaryPath.DIRECTORY_SEPARATOR;
        }

        $this->binaryPath = $binaryPath;

        return $this;
    }
```

The default `$binaryPath` is an empty string, calling `setBinaryPath()` with empty string results a new value: `/`.
The Optimizers are using `{$this->binaryPath}{$this->binaryName}` [in the `getCommand` function](https://github.com/spatie/image-optimizer/blob/master/src/Optimizers/BaseOptimizer.php#L54).

What do you think which make more sense: fixing this function call or adding empty string handling to `setBinaryPath()`?

(#106, #108, #110)
@nlemoine @freekmurze

**Comment by freekmurze** [comment]
I have no hard preference, but fixing it in `setBinaryPath` seems like a more robust fix.

Feel free to send a PR with tests.
